test/functional/sessions_controller_test.rb

```diff
@@ -78,6 +82,34 @@ class SessionsControllerTest < ActionController::TestCase
         end
       end
     end
+
+    context "when OTP is correct but session expired" do
+      setup do
+        @controller.session[:mfa_user] = @user.id
+        @controller.session[:mfa_expires_at] = 30.minutes.ago
```

Comment:
To make it more "real-world", you can utilize time traveling helpers (https://api.rubyonrails.org/classes/ActiveSupport/Testing/TimeHelpers.html#method-i-travel).

Reply:
TIL! Will make the change now, thanks :)